pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,443 @@
+		interfaceStore:       interfaceStore,
+		nodeConfig:           nodeConfig,
+		gwInformer:           gwInformer,
+		gwLister:             gwInformer.Lister(),
+		gwListerSynced:       gwInformer.Informer().HasSynced,
+		ciImportInformer:     ciImportInformer,
+		ciImportLister:       ciImportInformer.Lister(),
+		ciImportListerSynced: ciImportInformer.Informer().HasSynced,
+		queue:                workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "gatewayroute"),
+		installedCIImports:   cache.NewIndexer(clusterInfoImportKeyFunc, cache.Indexers{}),
+		namespace:            namespace,
+	}
+	controller.gwInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueueGateway(cur, false)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				controller.enqueueGateway(cur, false)
+			},
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueGateway(old, true)
+			},
+		},
+		resyncPeriod,
+	)
+	controller.ciImportInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueueClusterInfoImport(cur, false)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				controller.enqueueClusterInfoImport(cur, false)
+			},
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueClusterInfoImport(old, true)
+			},
+		},
+		resyncPeriod,
+	)
+	return controller
+}
+
+func clusterInfoImportKeyFunc(obj interface{}) (string, error) {
+	return obj.(*mcv1alpha1.ClusterInfoImport).Name, nil
+}
+
+func (c *MCRouteController) enqueueGateway(obj interface{}, isDelete bool) {
+	gw, isGW := obj.(*mcv1alpha1.Gateway)
+	if !isGW {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(nil, "Received unexpected object", "object", obj)
+			return
+		}
+		gw, ok = deletedState.Obj.(*mcv1alpha1.Gateway)
+		if !ok {
+			klog.ErrorS(nil, "DeletedFinalStateUnknown contains non-Gateway object", "object", deletedState.Obj)
+			return
+		}
+	}
+	if !isDelete {
+		if net.ParseIP(gw.InternalIP) == nil || net.ParseIP(gw.GatewayIP) == nil {
+			klog.ErrorS(nil, "No valid Internal IP or Gateway IP is found in Gateway", "gateway", gw.Namespace+"/"+gw.Name)
+			return
+		}
+	}
+	c.queue.Add("dummykey")
+}
+
+func (c *MCRouteController) enqueueClusterInfoImport(obj interface{}, isDelete bool) {
+	ciImp, isciImp := obj.(*mcv1alpha1.ClusterInfoImport)
+	if !isciImp {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(nil, "Received unexpected object", "object", obj)
+			return
+		}
+		ciImp, ok = deletedState.Obj.(*mcv1alpha1.ClusterInfoImport)
+		if !ok {
+			klog.ErrorS(nil, "DeletedFinalStateUnknown contains non-ClusterInfoImport object", "object", deletedState.Obj)
+			return
+		}
+	}
+
+	if !isDelete {
+		if len(ciImp.Spec.GatewayInfos) == 0 {
```

Comment:
Do you mean why to check this? it's similar as Gateway, we don't stop user to create or modify it manually, so check it here in case user changes it manually.